# ETL HECHOS

### IMPORTAR LIBRERIAS

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

### CARGAR DATASET

In [3]:
df_hechos = pd.read_excel(r'C:\PROYECTO-INDIVIDUAL-2\Data\xlsx\homicidios.xlsx', sheet_name='HECHOS')
df_hechos.head()

ID  N_VICTIMAS      FECHA  AAAA  MM  DD      HORA HH  \
0  2016-0001           1 2016-01-01  2016   1   1  04:00:00  4   
1  2016-0002           1 2016-01-02  2016   1   2  01:15:00  1   
2  2016-0003           1 2016-01-03  2016   1   3  07:00:00  7   
3  2016-0004           1 2016-01-10  2016   1  10  00:00:00  0   
4  2016-0005           1 2016-01-21  2016   1  21  05:20:00  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  ...  Altura  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA  ...     NaN   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ  ...     NaN   
2                         AV ENTRE RIOS 2034       AVENIDA  ...  2034.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA  ...     NaN   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA  ...     NaN   

                                 Cruce  \
0  FERNANDEZ DE LA CRUZ, F., GRAL. AV.   
1                  DE LOS CORRALES AV.   
2                                  NaN   
3             VILLEGAS, CONRADO, GRAL.   
4              SAENZ PE?A, LUIS, PRES.   

                               Dirección Normalizada COMUNA  \
0  PIEDRA BUENA AV. y FERNANDEZ DE LA CRUZ, F., G...      8   
1               PAZ, GRAL. AV. y DE LOS CORRALES AV.      9   
2                                ENTRE RIOS AV. 2034      1   
3          LARRAZABAL AV. y VILLEGAS, CONRADO, GRAL.      8   
4            SAN JUAN AV. y SAENZ PEÑA, LUIS, PRES.      1   

                                 XY (CABA)         pos x         pos y  \
0    Point (98896.78238426 93532.43437792)  -58.47533969  -34.68757022   
1    Point (95832.05571093 95505.41641999)  -58.50877521  -34.66977709   
2   Point (106684.29090040 99706.57687843)  -58.39040293  -34.63189362   
3    Point (99840.65224780 94269.16534422)  -58.46503904  -34.68092974   
4  Point (106980.32827929 100752.16915795)  -58.38718297  -34.62246630   

    PARTICIPANTES VICTIMA    ACUSADO  
0       MOTO-AUTO    MOTO       AUTO  
1  AUTO-PASAJEROS    AUTO  PASAJEROS  
2       MOTO-AUTO    MOTO       AUTO  
3         MOTO-SD    MOTO         SD  
4  MOTO-PASAJEROS    MOTO  PASAJEROS  

[5 rows x 21 columns]

### VISUALIZAMOS INFO

In [5]:
df_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 21 columns):
 #   Column                 Non-Null Count  Dtype         
---  ------                 --------------  -----         
 0   ID                     696 non-null    object        
 1   N_VICTIMAS             696 non-null    int64         
 2   FECHA                  696 non-null    datetime64[ns]
 3   AAAA                   696 non-null    int64         
 4   MM                     696 non-null    int64         
 5   DD                     696 non-null    int64         
 6   HORA                   696 non-null    object        
 7   HH                     696 non-null    object        
 8   LUGAR_DEL_HECHO        696 non-null    object        
 9   TIPO_DE_CALLE          696 non-null    object        
 10  Calle                  695 non-null    object        
 11  Altura                 129 non-null    float64       
 12  Cruce                  525 non-null    object        
 13  Direc

### ELIMINACION DE COLUMNAS QUE NO APORTAN INFORMACION
#### Como vemos con el head e info:
#### Las columnas XY(CABA) es lo mismo que pos x y pos y, por lo tanto podemos eliminarla, al igual que la direccón, el cruce y la calle, ya que solo utilizaremos la pos x, la pos y y el lugar del hecho para ubicarnos.
#### Participantes es una combinación de Victima y Acusado
#### Año, mes y día están incluidos en Fecha
#### Hora y HH (franja horaria) son similares, elegimos quedarnos con la segunda

In [7]:
columnas = ['Dirección Normalizada', 'Calle', 'Cruce',
            'AAAA', 'MM', 'DD', 'HORA', 'XY (CABA)', 'PARTICIPANTES']

df_hechos.drop(columns=columnas, inplace=True)

### VER VALORES NaN

In [9]:
tabla_hechos = []
tabla_victimas = []
tabla_hechos.append(pd.DataFrame({ 'NaN': df_hechos.isnull().sum()}))

for row in tabla_hechos:
    display(row)

NaN
ID                 0
N_VICTIMAS         0
FECHA              0
HH                 0
LUGAR_DEL_HECHO    0
TIPO_DE_CALLE      0
Altura           567
COMUNA             0
pos x              0
pos y              0
VICTIMA            0
ACUSADO            0

### ELIMINACION DE COLUMNA ALTURA

#### Como vemos Altura de las 696, tiene 567 vacias, por lo tanto la podemos descartar

In [10]:
df_hechos.drop(columns='Altura', inplace=True)

### VEMOS COMO VA QUEDANDO

In [11]:
df_hechos.head()

ID  N_VICTIMAS      FECHA HH  \
0  2016-0001           1 2016-01-01  4   
1  2016-0002           1 2016-01-02  1   
2  2016-0003           1 2016-01-03  7   
3  2016-0004           1 2016-01-10  0   
4  2016-0005           1 2016-01-21  5   

                             LUGAR_DEL_HECHO TIPO_DE_CALLE  COMUNA  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA       8   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ       9   
2                         AV ENTRE RIOS 2034       AVENIDA       1   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA       8   
4  AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA        AVENIDA       1   

          pos x         pos y VICTIMA    ACUSADO  
0  -58.47533969  -34.68757022    MOTO       AUTO  
1  -58.50877521  -34.66977709    AUTO  PASAJEROS  
2  -58.39040293  -34.63189362    MOTO       AUTO  
3  -58.46503904  -34.68092974    MOTO         SD  
4  -58.38718297  -34.62246630    MOTO  PASAJEROS

In [12]:
df_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 11 columns):
 #   Column           Non-Null Count  Dtype         
---  ------           --------------  -----         
 0   ID               696 non-null    object        
 1   N_VICTIMAS       696 non-null    int64         
 2   FECHA            696 non-null    datetime64[ns]
 3   HH               696 non-null    object        
 4   LUGAR_DEL_HECHO  696 non-null    object        
 5   TIPO_DE_CALLE    696 non-null    object        
 6   COMUNA           696 non-null    int64         
 7   pos x            696 non-null    object        
 8   pos y            696 non-null    object        
 9   VICTIMA          696 non-null    object        
 10  ACUSADO          696 non-null    object        
dtypes: datetime64[ns](1), int64(2), object(8)
memory usage: 59.9+ KB


### AHORA PASAREMOS A LIMPIAR NULOS, CAMBIAR TIPOS DE DATO Y RENOMBRAR COLUMNAS

#### La columna fecha está en el formato correcto

In [13]:
df_hechos['FECHA']

0     2016-01-01
1     2016-01-02
2     2016-01-03
3     2016-01-10
4     2016-01-21
         ...    
691   2021-12-13
692   2021-12-20
693   2021-12-30
694   2021-12-15
695   2021-11-18
Name: FECHA, Length: 696, dtype: datetime64[ns]

#### La columna HH tiene algunos valores 'sd' que representan los nulos por lo tanto hay que transformarlos

In [14]:
df_hechos['HH']

0       4
1       1
2       7
3       0
4       5
       ..
691    17
692     1
693     0
694    10
695     6
Name: HH, Length: 696, dtype: object

In [15]:
df_hechos['HH'].replace('sd', np.nan, inplace=True)
df_hechos['HH'] = pd.to_numeric(df_hechos['HH'], errors='coerce').astype('Int64')

#### Ahora pasaremos con las columnas que tendrían que ser tipo str, como LUGAR_DEL_HECHO, TIPO_DE_CALLE, VICTIMA y ACUSADO

In [16]:
columnas = ['LUGAR_DEL_HECHO', 'TIPO_DE_CALLE', 'VICTIMA', 'ACUSADO']
for columna in columnas:
    df_hechos[columna] = df_hechos[columna].str.strip()

#### Luego si vemos las comunas, en Bs As no existe la comuna 0, por lo tanto habría que poner como NaN, los valores 0 de esa columna

In [17]:
df_hechos['COMUNA'].replace(0, np.nan, inplace=True)
df_hechos['COMUNA']

C:\Users\augus\AppData\Local\Temp\ipykernel_172\3315344530.py:1: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  df_hechos['COMUNA'].replace(0, np.nan, inplace=True)


0       8.0
1       9.0
2       1.0
3       8.0
4       1.0
       ... 
691     7.0
692     9.0
693    11.0
694     9.0
695     1.0
Name: COMUNA, Length: 696, dtype: float64

#### Ahora pasaremos a renombrar las columnas para facilitar su uso

In [18]:
Actuales = df_hechos.columns
Nuevas = ['id_hecho', 'cantidad_victimas', 'fecha', 'hora', 'lugar', 
       'tipo_de_calle', 'comuna', 'longitud', 'latitud', 'vehiculo_victima', 'vehiculo_acusado']

columnas = {Actuales: Nuevas for (Actuales, Nuevas) in zip(Actuales, Nuevas)}
df_hechos.rename(columns=columnas, inplace=True)

### Ahora vamos a ver como quedó todo y guardar los df limpios

In [19]:
df_hechos.head()

id_hecho  cantidad_victimas      fecha  hora  \
0  2016-0001                  1 2016-01-01     4   
1  2016-0002                  1 2016-01-02     1   
2  2016-0003                  1 2016-01-03     7   
3  2016-0004                  1 2016-01-10     0   
4  2016-0005                  1 2016-01-21     5   

                                       lugar tipo_de_calle  comuna  \
0  AV PIEDRA BUENA Y AV FERNANDEZ DE LA CRUZ       AVENIDA     8.0   
1          AV GRAL PAZ  Y AV DE LOS CORRALES      GRAL PAZ     9.0   
2                         AV ENTRE RIOS 2034       AVENIDA     1.0   
3      AV LARRAZABAL Y GRAL VILLEGAS CONRADO       AVENIDA     8.0   
4   AV SAN JUAN Y PRESIDENTE LUIS SAENZ PEÑA       AVENIDA     1.0   

       longitud       latitud vehiculo_victima vehiculo_acusado  
0  -58.47533969  -34.68757022             MOTO             AUTO  
1  -58.50877521  -34.66977709             AUTO        PASAJEROS  
2  -58.39040293  -34.63189362             MOTO             AUTO  
3  -58.46503904  -34.68092974             MOTO               SD  
4  -58.38718297  -34.62246630             MOTO        PASAJEROS

In [20]:
df_hechos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 696 entries, 0 to 695
Data columns (total 11 columns):
 #   Column             Non-Null Count  Dtype         
---  ------             --------------  -----         
 0   id_hecho           696 non-null    object        
 1   cantidad_victimas  696 non-null    int64         
 2   fecha              696 non-null    datetime64[ns]
 3   hora               695 non-null    Int64         
 4   lugar              696 non-null    object        
 5   tipo_de_calle      696 non-null    object        
 6   comuna             694 non-null    float64       
 7   longitud           696 non-null    object        
 8   latitud            696 non-null    object        
 9   vehiculo_victima   696 non-null    object        
 10  vehiculo_acusado   696 non-null    object        
dtypes: Int64(1), datetime64[ns](1), float64(1), int64(1), object(7)
memory usage: 60.6+ KB


In [21]:
df_hechos.to_parquet(r'C:\PROYECTO-INDIVIDUAL-2\Data\parquet\hechos.parquet', index=False)